azavea/climate-change-lab
# Download chart as image

## Overview

Add button to chart toolbar to download chart as an image (PNG).


### Demo

New toolbar button:
![image](https://cloud.githubusercontent.com/assets/960264/21899842/ca787bc0-d8bf-11e6-8f88-b5d32fa9eedf.png)


Clicking leftmost toolbar button downloads this:
![extreme_cold_events](https://cloud.githubusercontent.com/assets/960264/21899808/aa19ed82-d8bf-11e6-856e-a93d5f836d10.png)

File name is the indicator name.


### Notes

The SVG to PNG conversion probably doesn't work on IE; this maybe could be accomplished by writing the SVG to canvas first.

Attempted to also offer option to share PNG on social media, but did not follow through, as it would be complicated by requiring a server-side component for authentication with the social media service, and could not be accomplished with standard social sharing widgets. Image would need to be uploaded first, then the returned identifier for the media object passed in to the status post. We would also have to put together our own edit/preview pane. I left in the intermediate commits for adding a social media icon popover to the chart toolbar button, in case that comes in handy later. Largely based on [this demo](http://codepen.io/labnol/pen/CenhB), with various CSS fixes.

For Twitter posting, [this endpoint](https://dev.twitter.com/rest/reference/post/statuses/update_with_media), though deprecated, would be the most straightforward to use for the post with image.


## Testing Instructions

 * Will need to `npm install` to pick up SVG-to-PNG converter package
 * Each chart should have new download toolbar button
 * Clicking image download button should immediately download appropriate chart, with CSS reflected
